helm/charts/oathkeeper/values.yaml

```diff
@@ -111,3 +111,19 @@ tolerations: []
 
 # Configure node affinity
 affinity: {}
+
+# Extra config files that will be put to the volume at "/etc/config".
+# For example, you can define multiple additional access rules files.
+extraConfigFiles: []
```

Comment:
Would it make more sense if we changed `accessRules` from `Rule (Object)` to:

```yaml
oathkeeper:
  accessRules:
    - file-a.yaml
    - file-b.yaml
```

?

Reply:
I like the idea. But this is not enough, because we need both file name and file contents. 

Option 1)

```yaml
oathkeeper:
  accessRules:
    rules.yaml: |-
      - id: foo
```

But in this case you need to be careful when using `--set-file` because periods must be escaped: `--set-file "oathkeeper.accessRules.foo\.json=./helm/charts/oathkeeper/demo/access-rules.json"`

Option 2)

```yaml
oathkeeper:
  accessRules:
    - filename: rules.yaml
      contents: |-
        -id: foo
```

This is more tedious but no extra escaping required.

---

Also, I suggest introducing another parameter for `values` file where you can write rules not as a string, but as yaml (recently I added yaml support for rules here: https://github.com/ory/oathkeeper/pull/213), for example:

```yaml
oathkeeper:
  accessRules:
    - id: foo
      authorizer:
        handler: noop
```

I see this is by far more convenient than yaml strings with `|-`.

So we can have 2 paramerers: one for inline yaml and another for files, e.g.

```yaml
oathkeeper:
  accessRules:
    - id: foo
      authorizer:
        handler: noop
  accessRulesFiles:
      rules.yaml: |-
        - id: foo
```

@aeneasr  What do you think?

Reply:
> I like the idea. But this is not enough, because we need both file name and file contents.

Can we not extract the filename from above? E.g. with a hash?

```
oathkeeper:
  accessRules:
    - file-a.yaml
    - file-b.yaml
```

Results in contents from `file-a.yaml` with filename `md5("file-a.yml")` and `file-b.yaml` with filename `md5("file-b.yml")`. What do you think?

> Also, I suggest introducing another parameter for values file where you can write rules not as a string, but as yaml (recently I added yaml support for rules here:

I think this could be a nice addition later on (e.g. `accessRuleInline`) but I'd leave it out of this PR for now and see if there are use cases where people demand this :)

Reply:
> Can we not extract the filename from above? E.g. with a hash?

The main problem is how to get the file contents, not the file name. Helm is very limited in reading files by name reference, so we definitely should support `--set-file` instead.

Reply:
> The main problem is how to get the file contents, not the file name. Helm is very limited in reading files by name reference, so we definitely should support --set-file instead.

Ohh I see - that makes absolute sense. How about Option 3 then:

```yaml
oathkeeper:
  accessRules:
    - rules: |-
      - id: rule-1
        authenticator: ...
      - id: rule-2
        authenticator: ...
```

where the filename is generated by (`md5(accessRules.[*].rules)`)?

Reply:
So this is Option 1 basically, not the 3rd one (still using stirngs via `|-`).

In this case, why do we need md5? `rules` word in your example can be a filename:

```yaml
oathkeeper:
  accessRules:
    - rules.yaml: |-
      - id: rule-1
        authenticator: ...
      - id: rule-2
        authenticator: ...
```

Reply:
Good point, I was thinking that we might want to have other options too, such as:

```
oathkeeper:
  accessRules:
    - onlyForAnntation: 
       - annotation/foo # this access rule would only be deployed on pods with this annotation
       rules: |-
       ...
```

But that's maybe a stretch, it would however allow for easy changes later on. Adding such a value would be difficult in the

```
oathkeeper:
  accessRules:
    - rules.yaml: |-
      - id: rule-1
        authenticator: ...
      - id: rule-2
        authenticator: ...
```

model because we don't know what the file name is and what a config option is.

Reply:
Personally, I think an example with annotations adds unnecessary complexity (the same issue with `md5` stuff instead of explicit filename).

> Adding such a value would be difficult in the ... model because we don't know what the file name is and what a config option is.

I'm not sure I understand the difficulty, isn't it simply `--set-file "oathkeeper.accessRules[0].rules\.yaml=somefile.yaml" `?

---

Let's sum up what we have. I propose one of the following:

1. Option with "string" config formatting, one of two variants:

    1. First

        ```yaml
        oathkeeper:
        accessRules:
          rules.yaml: |-
            - id: foo
        ```

        How to configure:

        1. Directly in the values.yaml file

        1. Using `--set-file "oathkeeper.accessRules[0].rules\.yaml=somefile.yaml"`

        Examples in other charts:

        - [rabbitmq](https://github.com/helm/charts/blob/master/stable/rabbitmq-ha/values.yaml#L32)
        - [fluentd](https://github.com/helm/charts/blob/master/stable/fluentd/values.yaml#L76)

    1. Second

        ```yaml
        oathkeeper:
        accessRules:
          - filename: rules.yaml
            contents: |-
              -id: foo
        ```

        How to configure:

        1. Directly in the values.yaml file

        1. Using `--set "oathkeeper.accessRules[0].filename=rules.yaml"` and `--set-file "oathkeeper.accessRules[0].contents=somefile.yaml"`

        Examples:

        - [coredns](https://github.com/helm/charts/blob/master/stable/coredns/values.yaml#L120) zone files

2. Option with "inline" yaml config formatting, similar to how `oathkeeper.config` section:

    ```yaml
    oathkeeper:
    accessRules:
      - id: foo
        authorizer:
          handler: noop
    ```

    This approach used widely in helm charts when the application supports yaml (or compatible) configuration as it provides native configuration using both in values.yaml file and using `--set` arguments.

    Examples:

    - [prometheus](https://github.com/helm/charts/blob/master/stable/prometheus/values.yaml#L981)
    - [dex](https://github.com/helm/charts/blob/master/stable/dex/values.yaml#L134)
    - [grafana](https://github.com/helm/charts/blob/master/stable/grafana/values.yaml#L333)
    - [mongodb](https://github.com/helm/charts/blob/master/stable/mongodb/values.yaml#L332)
    - [influxdb](https://github.com/helm/charts/blob/master/stable/influxdb/values.yaml#L135)

So what should we stick into?

Reply:
Sorry for my slow response, there was so much stuff going on in the last couple of days.

Thank you for the great sum up. In retrospective, I think all 3 approaches are great. I think we should go with the approach that is most common, so here's my final proposal.

We change `oathkeeper.accessRules` from a single file to an array of files with yaml:

```yaml
oathkeeper:
accessRules:
  rules.yaml:
    - id: foo
      authorizer:
        handler: noop
```

Does that sound good to you?